File: r/fib.r
fib <- function(n) {
  if (n < 2) n else Recall(n - 1) + Recall(n - 2)
}

fibLinear <- function(n) {
  fibs <- c(0, 1)
  while (n > 0) {
    temp <- fibs[2]
    fibs[2] <- sum(fibs)
    fibs[1] <- temp
    n <- n - 1
  }
  fibs[1]
}

fibFormula <- function(n) {
  round(((5 ^ 0.5 + 1) / 2) ^ n / 5 ^ 0.5)
}

fibTailRecursive <- function(n, prevFib = 0, fib = 1) {
  if (n == 0) prevFib else Recall(n - 1, fib, prevFib + fib)
}

args = commandArgs(trailingOnly=TRUE)
input = if (length(args) > 0) as.numeric(args[1]) else 29
print(fibFormula(input))

File: julia/fib.jl
fib(n) = n < 2 ? n : fib(n-1) + fib(n-2)

fibTailRecursive(a, b=0, c=1) = a === 0 ? b : fibTailRecursive(a - 1, c, b + c)

fibFormula(n) = round(((5 ^ .5 + 1) / 2) ^ n / (5 ^ .5))

function fibLinear(n)
  x,y = (0,1)
  for i = 1:n
    x,y = (y, x+y)
  end
  return x
end

function fibMatrix(n)
  prevMatrix = BigInt[1 1; 1 0]
  matrix = prevMatrix ^ n
  matrix[2, 1]
end

input = length(ARGS) === 0 ? 29 : parse(Int, ARGS[1])
println(fibMatrix(input))
